tests/test_repository.py

```diff
@@ -135,9 +121,71 @@ def test_package_is_uploaded_200s_with_no_releases():
     assert repo.package_is_uploaded(package) is False
 
 
+def test_package_is_uploaded_with_releases_using_cache(repo):
+    """Test that a package upload succeeds without bypassing cache"""
+    repo._releases_json_data = {"fake": {"0.1": [{"filename": "fake.whl"}]}}
+    package = pretend.stub(
+        safe_name="fake", basefilename="fake.whl", metadata=pretend.stub(version="0.1"),
+    )
+
+    assert repo.package_is_uploaded(package) is True
+
+
+def test_package_is_uploaded_with_releases_not_using_cache(repo):
+    """Test that a package upload succeeds bypassing cache"""
+    repo.session = pretend.stub(
+        get=lambda url, headers: response_with(
+            status_code=200,
+            _content=b'{"releases": {"0.1": [{"filename": "fake.whl"}]}}',
+            _content_consumed=True,
+        ),
+    )
+    package = pretend.stub(
+        safe_name="fake", basefilename="fake.whl", metadata=pretend.stub(version="0.1"),
+    )
+
+    assert repo.package_is_uploaded(package, bypass_cache=True) is True
+
+
+def test_package_is_uploaded_different_filenames(repo):
+    """Test that a package upload fails as safe name and basefilename differ"""
+    repo.session = pretend.stub(
+        get=lambda url, headers: response_with(
+            status_code=200,
+            _content=b'{"releases": {"0.1": [{"filename": "fake.whl"}]}}',
+            _content_consumed=True,
+        ),
+    )
+    package = pretend.stub(
+        safe_name="fake", basefilename="foo.whl", metadata=pretend.stub(version="0.1"),
+    )
+
+    assert repo.package_is_uploaded(package) is False
+
+
+def test_package_is_uploaded_incorrect_repo_url(repo):
+    """Test that a package upload fails on providing wrong repo url"""
+    repo.url = "https://bad.repo.com/legacy"
+    assert repo.package_is_uploaded(None) is False
```

Comment:
I probably would not use `repo` here, but instead explicitly construct a `Repository` instance with a bad URL.

I'd also probably move this to the bottom of the file, to keep it distinct from all of the test that use a valid repo URL.

Reply:
Fixed in latest commit